### PR TITLE
Add support for returning multiple messages from `gr.ChatInterface` chat function

### PR DESCRIPTION
**Pull request by abidlabs**
This PR firstly allows returning a `list` from `gr.ChatInterface`, which gets rendered as multiple assistant messages. Secondly, it allows files to be returned from a chat function just by returning a `tuple`, which is not only simpler but also more consistent with how files are actually stored in the history. Also improves test coverage for API use.

See `demo/chatinterface_echo_multimodal/run.py` that demonstrates both of these changes.

### CODE REVIEW

**Review by aliabd — COMMENTED** [review]
The changes lgtm @abidlabs. theres just two things: 
- I think you need to fix the end of the docstring for the `fn` parameter to reflect the changes. 
- the $code_chatinterface_options breaks the page rendering because it contains 

```
example_code = """
Here's the code I generated:

```python
def greet(x):
    return f"Hello, {x}!"
\```

Is this correct?
"""
```

the ```python is breaking the page. I've seen this before and because i didn't want to change the demo, my suggestion is to instead copy/paste the code here and just remove the backticks/code syntax.

**Comment by abidlabs** [comment]
Thanks @aliabd, addressed both of the comments above

**Review by freddyaboulton — COMMENTED** [review]
Nice PR @abidlabs ! 

One comment is that instead of supporting tuples as files we should point developers towards using a gradio component like gr.File or gr.Image.

I tried converting the demo you wrote to the tuples case and it's resulting in an error because it's treating `(None, "You wrote...")` as a file. I guess we could support returning a list of lists but that conflicts with the parameter name `type="tuples"`.

```python
import gradio as gr

def echo_multimodal(message, history):
    response = []
    response.append((None, "You wrote: '" + message["text"] + "' and uploaded:"))
    for file in message.get("files", []):
        response.append((None, gr.File(value=file)))
    return response

demo = gr.ChatInterface(
    echo_multimodal,
    multimodal=True,
    type="tuples",
    textbox=gr.MultimodalTextbox(file_count="multiple"),
)

if __name__ == "__main__":
    demo.launch()
```

So I think it would break the ambiguity to use a gradio component.

**Comment by abidlabs** [comment]
Ah I see okay yeah let's avoid that for now. I'll revert the docs changes around the file changes

**Comment by abidlabs** [comment]
Should be ok now @freddyaboulton

File: gradio/chat_interface.py [hunk 1]
```diff
@@ -618,13 +619,16 @@ async def _submit_fn(
...
             )
-        if isinstance(response, tuple):
+        if self.additional_outputs:
             response, *additional_outputs = response
         else:
             additional_outputs = None
         history = self._append_message_to_history(message, history, "user")
-        response_ = self.response_as_dict(response)
-        history = self._append_message_to_history(response_, history, "assistant")  # type: ignore
+        response_ = [response] if not isinstance(response, list) else response
```
Comment (freddyaboulton) [inline gradio/chat_interface.py:627]: We should add this logic to `_stream_fn` to support streaming. Otherwise LGTM!

Comment (abidlabs) [reply gradio/chat_interface.py:627]: Nice catch thank you!

**Comment by abidlabs** [comment]
Thanks for the feedback @freddyaboulton @aliabd!

**Pull request merged**